### PR TITLE
57 Update fit models script with q0.5 model and add function to check for model drift 

### LINKED ISSUE #56: fit_cost_prediction_intervals.py: Train model for median cost

### LINKED ISSUE #57: Check for drift

### PR DESCRIPTION
**Pull request by sofiapinto**
---

# Description

In this PR we:
- Integrate a 50th percentile model in the `fit_cost_prediction_intervals.py` script
- Add a function to check for model drift

Fixes #56 #57 

# Instructions for Reviewer

Hey @caldwellst , thanks for taking the time to review this!

Could you please review this script `asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py`?

If you run this now the drift analysis won't work as we don't have previous results stored to make the comparison, but this will be possible after rerunning this with the newest quarter data. Would you add/change anything to the drift analysis?

You can run this on test mode with:
`python asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py --lower_quantile 0.1 --upper_quantile 0.9 --test True`

# Checklist:

- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py [hunk 1]
```diff
@@ -103,89 +174,189 @@ def set_up_pipeline(quantile: float, model_bound: str) -> Pipeline:
...
+        assess_drift = False
+        models_performance = dict()
+
+    # Append todays_model_performance to models_performance
+    models_performance[today_date] = todays_model_performance
+
+    if not test_mode:
+        # Save updated models_performance to s3
+        models_performance.to_json(
+            f"s3://asf-hp-cost-estimator-model/outputs/model_performance.json",
```
Comment (caldwellst) [inline asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py:322]: Might be worth just injecting a date string in here so we can read in the latest, but also have a record of model performance across time.

Comment (caldwellst) [reply asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py:322]: Or just put into a single data frame actually. We want to have these automated checks, but even better is also just having a plot of time series of the metrics across time. We might be getting 8% drift in the metrics every quarter, but never flagging, and end up after a year with a significantly degraded model (doubt it but might be worth checking).

File: asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py [hunk 2]
```diff
@@ -103,89 +174,189 @@ def set_up_pipeline(quantile: float, model_bound: str) -> Pipeline:
...
+        )
+
+        previous_model_performance = models_performance[most_recent_folder]
+        for df in ["full_dataset", "latest_quarter"]:
+            logging.info(f"Comparing performance on {df}...")
+            today_df_metrics = todays_model_performance[df]
+            previous_df_metrics = previous_model_performance[df]
+            for key in today_df_metrics.keys():
+                # Check if todays_model_performance is more than 10% worse than previous_model_performance
+                if today_df_metrics[key] > previous_df_metrics[key] * 1.1:
```
Comment (caldwellst) [inline asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py:339]: Hmm, only exception is `coverage` where we are technically looking for divergence in either direction.

**Review by caldwellst — CHANGES_REQUESTED** [review]
Didn't go into too much detail on reviewing the exact script. The one comment I put in is making sure that we are saving this data across time, which I think would be really valuable, and enable us to visualise how performance has shifted.

However, thinking about the problem, I'm wondering if looking at general model performance as measured on the final fit is what we want to be recording. Rather, I think we should be tracking performance on the hold out test from the latest period. Checks should focus on whether we're seeing degradation at this specific latest period, which is actually what the model is used for. Don't really care as much about performance in previous time periods, and also as is, we are capturing performance metrics on the same data the model was trained on.

I don't want to lose too much time on this, and is not a priority for going live. But is there some simple shifts to the `evaluate_and_log_metrics()` function that automatically saves out the performance measurement. Would be really simple change there, and then we'll have performance saved out on the cross-validation set, the test set, and the validation set. It would also mean that you don't have to compute any new metrics when checking for drift, in pseudocode you could simply do:

- Read in all metrics JSON files into a data frame, parsing dates in the filename to a column
- Compare the latest metric (which would be from the current run) with the previous metric
- If you want, plot a time series as a graph somewhere

Again, I think this can go on the backlog compared to going live with this model.
